source/task_scheduler.c

```diff
@@ -0,0 +1,123 @@
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/task_scheduler.h>
+#include <aws/common/thread.h>
+#include <aws/common/priority_queue.h>
+
+static const size_t DEFAULT_QUEUE_SIZE = 7;
+
+struct task_container {
+    uint64_t timestamp;
+    struct aws_task task;
+};
+
+static int compare_timestamps(const void *a, const void *b) {
+    uint64_t a_time = ((struct task_container*)a)->timestamp;
+    uint64_t b_time = ((struct task_container*)b)->timestamp;
+    return a_time > b_time; // min-heap
+}
+
+int aws_task_scheduler_init_default(struct aws_task_scheduler *scheduler, aws_task_scheduler_clock clock, 
+        struct aws_allocator *alloc) {
+    scheduler->alloc = alloc;
+    scheduler->clock = clock;
+    scheduler->owning_thread = aws_thread_current_thread_id();
+    return aws_priority_queue_dynamic_init(&scheduler->queue, alloc, DEFAULT_QUEUE_SIZE, sizeof(struct task_container), 
+            &compare_timestamps);
+}
+
+void aws_task_scheduler_init_with_queue(struct aws_task_scheduler *scheduler,
+        struct aws_priority_queue *queue, aws_task_scheduler_clock clock) {
+
+    scheduler->alloc = NULL;
+    scheduler->clock = clock;
+/* I feel icky about copying the queue. If we don't copy it, then who owns the queue? this is making me question this whole API. Need feedback */
+    scheduler->queue = *queue;
+    scheduler->owning_thread = aws_thread_current_thread_id();
+}
+
+void aws_task_scheduler_clean_up(struct aws_task_scheduler *scheduler) {
+     /* do we want to add a xthread check here? what happens if another thread attempts to clean up? */
+    aws_priority_queue_clean_up(&scheduler->queue);
+    scheduler->alloc = NULL;
+    scheduler->clock = NULL;
+    scheduler->owning_thread = 0;
+}
+
+int aws_task_scheduler_next_task(struct aws_task_scheduler *scheduler, 
+        struct aws_task *task, uint64_t *next_run_time) {
+
+    if (aws_thread_current_thread_id() != scheduler->owning_thread) {
+        return aws_raise_error(AWS_ERROR_TASK_SCHEDULER_ILLEGAL_XTHREAD_ACCESS);
+    }
+
+    if (next_run_time) {
+        *next_run_time = 0;
+    }
+
+    struct task_container *possible_task;
+    if(AWS_OP_SUCCESS == aws_priority_queue_top(&scheduler->queue, (void **)&possible_task)) {
+
+        if (next_run_time) {
+            *next_run_time = possible_task->timestamp;
+        }
+
+        uint64_t now;
+        if(scheduler->clock(&now)) {
+            return AWS_OP_ERR;
+        }
+
+        if (possible_task->timestamp <= now) {
+            struct task_container container;
+            if(aws_priority_queue_pop(&scheduler->queue, (void *)&container)) {
+                return AWS_OP_ERR;
+            }
+            *task = container.task;
+            return AWS_OP_SUCCESS;
```

Comment:
next_run_time is incorrect at this point, since the 'next run time' would be that of the new top of the heap.